### PR TITLE
Changes to deploy to AKS in non prod

### PR DESCRIPTION
**Pull request by adusumillipraveen**
Notes:
* Deploys apps to AAT AKS controlled through config
* Docker builds on master are tagged as 'staging' for AAT staging and tagged as latest along with prod after tests are passed.
* Optionally disable deploying to AAT ASEs with legacy deployment flags.
* May look scary due to white space changes .

### CODE REVIEW

File: src/uk/gov/hmcts/contino/AppPipelineDsl.groovy [hunk 1]
```diff
@@ -79,4 +79,8 @@ class AppPipelineDsl extends CommonPipelineDsl implements Serializable {
     config.vaultEnvironmentOverrides = vaultOverrides
   }
 
+  void installChartsToNonProd() {
```
Comment (timja) [inline src/uk/gov/hmcts/contino/AppPipelineDsl.groovy]:
```suggestion
  void enableAksStagingDeployment() {
```

File: vars/sectionDeployToDeploymentTarget.groovy [hunk 2]
```diff
@@ -63,136 +64,137 @@ def call(params) {
     }
   }
 
-  // merge env and deployment target tf output using some groovy magic
-  def mergedTfOutput
-  if (tfOutput) {
-    envTfOutput.properties.each {
-      tfOutput.metaClass[it.key] = it.value
+  if(config.legacyDeployment || !(new Environment(env).nonProdName == environment)) {
```
Comment (timja) [inline vars/sectionDeployToDeploymentTarget.groovy]:
```suggestion
  if (config.legacyDeployment || !(new Environment(env).nonProdName == environment)) {
```

File: vars/sectionPromoteBuildToStage.groovy [hunk 3]
```diff
@@ -44,6 +44,9 @@ def call(params) {
 
         pcr.callAround("${deploymentStage.label}:promotion") {
           acr.retagForStage(deploymentStage, dockerImage)
+          if(DockerImage.DeploymentStage.PROD == deploymentStage) {
```
Comment (timja) [inline vars/sectionPromoteBuildToStage.groovy]:
```suggestion
          if (DockerImage.DeploymentStage.PROD == deploymentStage) {
```

File: vars/withPipeline.groovy [hunk 4]
```diff
@@ -71,28 +71,20 @@ def call(type, String product, String component, Closure body) {
...
-          pipelineCallbacksRunner: callbacksRunner,
-          pipelineType: pipelineType,
-          subscription: subscription.nonProdName,
-          environment: environment.previewName,
-          product: product,
-          component: component
-        )
-
-        onMaster {
+        onPR{
```
Comment (timja) [inline vars/withPipeline.groovy]:
```suggestion
        onPR {
```

File: vars/sectionCI.groovy [hunk 5]
```diff
@@ -35,35 +35,42 @@ def call(params) {
...
+          stage('Deploy to AKS') {
+            pcr.callAround('aksdeploy') {
+              timeoutWithMsg(time: 15, unit: 'MINUTES', action: 'Deploy to AKS') {
+                onPR {
                   deploymentNumber = githubCreateDeployment()
+                }
 
-                  aksUrl = aksDeploy(dockerImage, params)
-                  log.info("deployed component URL: ${aksUrl}")
+                aksUrl = aksDeploy(dockerImage, params)
```
Comment (timja) [inline vars/sectionCI.groovy]: we need to think about binning this soon

File: vars/sectionCI.groovy [hunk 6]
```diff
@@ -73,39 +80,66 @@ def call(params) {
...
-      if (config.deployToAKS || config.installCharts) {
-        withSubscription(subscription) {
-          withTeamSecrets(config, environment) {
-            stage("Smoke Test - AKS") {
+    if (config.deployToAKS || config.installCharts) {
+      withSubscription(subscription) {
+        withTeamSecrets(config, environment) {
+          stage("Smoke Test - AKS ${environment}") {
+            testEnv(aksUrl) {
+              pcr.callAround("smoketest:aks") {
```
Comment (timja) [inline vars/sectionCI.groovy]: needs environment in it

Comment (timja) [reply vars/sectionCI.groovy]:
```suggestion
              pcr.callAround("smoketest:${environment}-staging") {
```

File: vars/sectionCI.groovy [hunk 7]
```diff
@@ -35,35 +35,42 @@ def call(params) {
...
           }
-        } else if (config.installCharts) {
-          withTeamSecrets(config, environment) {
-            stage('Install Charts to AKS') {
-              pcr.callAround('akschartsinstall') {
-                timeoutWithMsg(time: 15, unit: 'MINUTES', action: 'Install Charts to AKS') {
+        }
+      } else if (config.installCharts) {
+        withTeamSecrets(config, environment) {
+          stage('Install Charts to AKS ${environment}') {
```
Comment (timja) [inline vars/sectionCI.groovy]:
```suggestion
          stage('AKS deploy - ${environment}') {
```

File: vars/sectionCI.groovy [hunk 8]
```diff
@@ -73,39 +80,66 @@ def call(params) {
...
+                    builder.performanceTest()
+                    publishPerformanceReports(this, params)
+                  }
+                }
+              }
+            }
+          }
+          onMaster {
+            if (config.crossBrowserTest) {
+              stage("CrossBrowser Test - AKS ${environment}") {
```
Comment (timja) [inline vars/sectionCI.groovy]:
```suggestion
              stage("Cross browser Test - AKS ${environment}") {
```

File: vars/sectionDeployToDeploymentTarget.groovy [hunk 9]
```diff
@@ -63,136 +64,137 @@ def call(params) {
     }
   }
 
-  // merge env and deployment target tf output using some groovy magic
-  def mergedTfOutput
-  if (tfOutput) {
-    envTfOutput.properties.each {
-      tfOutput.metaClass[it.key] = it.value
+  if (config.legacyDeployment || !(new Environment(env).nonProdName == environment)) {
```
Comment (timja) [inline vars/sectionDeployToDeploymentTarget.groovy]: maybe a comment

File: vars/withPipeline.groovy [hunk 10]
```diff
@@ -103,6 +95,18 @@ def call(type, String product, String component, Closure body) {
             product: product,
             component: component)
 
+          if (pipelineConfig.aksStagingDeployment) {
+            sectionCI(
```
Comment (timja) [inline vars/withPipeline.groovy]: this seems a bit hacky re-using sectionCI

CI should be for continuous integration not for deployment

Maybe get this PR in and then look at refactoring it?
Not sure

File: vars/sectionCI.groovy (hunk 5 above)
Comment (adusumillipraveen) [reply vars/sectionCI.groovy]: yeah, only few seem to be left in migration :
https://tools.hmcts.net/confluence/display/CNP/Helm+Charts+Migration+Progress

File: vars/withPipeline.groovy (hunk 10 above)
Comment (adusumillipraveen) [reply vars/withPipeline.groovy]: Renamed it .

File: vars/sectionCI.groovy (hunk 6 above)
Comment (adusumillipraveen) [reply vars/sectionCI.groovy]: Do we need -staging in here?

Comment (timja) [reply vars/sectionCI.groovy]: probably ok without

**Pull request merged**